libobs-opengl/gl-x11-egl.c

```diff
@@ -0,0 +1,671 @@
+/******************************************************************************
+    Copyright (C) 2019 by Ivan Avdeev <me@w23.ru>
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 2 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+******************************************************************************/
+
+/* GL context initialization using EGL instead of GLX
+ * This is needed in order to allow GL interaction with DMA-BUF machinery.
+ * Which is essential for improved and more performant screen grabbing and
+ * VA-API feeding techniques.
+ *
+ * See FIXME dma-buf input source
+ *
+ * Note: most of x11-related functionality was taken from gl-x11.c
+ */
+
+#include <X11/Xlib.h>
+#include <X11/Xlib-xcb.h>
+
+#include <xcb/xcb.h>
+
+#include <stdio.h>
+
+#include "gl-subsystem.h"
+
+#include <glad/glad_egl.h>
+
+/* FIXME glad */
+typedef EGLDisplay(EGLAPIENTRYP PFNEGLGETPLATFORMDISPLAYEXTPROC)(
+	EGLenum platform, void *native_display, const EGLint *attrib_list);
+
+/* clang-format off */
+static const int ctx_attribs[] = {
+#ifdef _DEBUG
+	EGL_CONTEXT_OPENGL_DEBUG, EGL_TRUE,
+#endif
+	EGL_CONTEXT_OPENGL_PROFILE_MASK, EGL_CONTEXT_OPENGL_CORE_PROFILE_BIT,
+	EGL_CONTEXT_MAJOR_VERSION, 3, EGL_CONTEXT_MINOR_VERSION, 2,
```

Comment:
Recently changed to require 3.3

Reply:
> > It likely doesn't work on nvidia.
> 
> On my system I see (using obs-drmsend as root and setuid)
> 
> ```
> obs-drmsend: Opening card /dev/dri/card0
> Cannot get drm planes: 95: Operation not supported
> ```

If you have nvidia card with binary drivers, this method should not work at all. I don't have NV hw myself here easily available, but from what I read, binary drivers don't provide DMA-BUF interface, and use EGLStreams instead. Figuring out whether it's possible to use that for plumbing display framebuffers to GL textures requires research.

Reply:
Will change to 3.3, thanks!

(Not sure why previous reply ended up here too, I can't into computers)